drivers/usb/device/usb_dc_it82xx2.c

```diff
@@ -0,0 +1,1564 @@
+ * the entrance of configuring the EPN0N1 Extended Ctrl Registers.
+ *
+ * The variable set_clr determines if we set/clear the corresponding bit.
+ */
+static void epn0n1_ext_ctrl_cfg(uint8_t reg_idx, uint8_t bit_mask,
+			bool set_clr)
+{
+	(set_clr) ? (epn0n1_ext_ctrl[reg_idx] |= bit_mask) :
+		(epn0n1_ext_ctrl[reg_idx] &= ~(bit_mask));
+}
+
+static void epn0n1_ext_ctrl_cfg_seq_inv(uint8_t reg_idx, uint8_t bit_mask,
+			bool set_clr)
+{
+	bool check = (set_clr) ?
+		(epn0n1_ext_ctrl[reg_idx] & EPN0_OUTDATA_SEQ) :
+		(epn0n1_ext_ctrl[reg_idx] & EPN1_OUTDATA_SEQ);
+
+	(check) ? (epn0n1_ext_ctrl[reg_idx] &= ~(bit_mask)) :
+		(epn0n1_ext_ctrl[reg_idx] |= bit_mask);
+}
+
+/*
+ * Function epn01n1_check_stall():
+ *
+ * Return the status of STALL bit in the EPN0N1 Extend Control Registers
+ */
+static bool epn01n1_check_stall(uint8_t reg_idx, uint8_t bit_mask)
+{
+	return !!(epn0n1_ext_ctrl[reg_idx] & bit_mask);
+}
+
+/*
+ * The following functions provide the entrance of configuring the
+ * EPN Extended Ctrl Registers.
+ */
+static void epn_ext_ctrl_cfg1(uint8_t reg_idx, uint8_t bit_mask, bool set_clr)
+{
+	(set_clr) ? (epn_ext_ctrl[reg_idx].EPN_EXT_CTRL1 |= bit_mask) :
+		(epn_ext_ctrl[reg_idx].EPN_EXT_CTRL1 &= ~(bit_mask));
+}
+
+static void epn_ext_ctrl_cfg2(uint8_t reg_idx, uint8_t bit_mask, bool set_clr)
+{
+	(set_clr) ? (epn_ext_ctrl[reg_idx].EPN_EXT_CTRL2 |= bit_mask) :
+		(epn_ext_ctrl[reg_idx].EPN_EXT_CTRL2 &= ~(bit_mask));
+}
+
+/* From 98h to 9Dh, the EP45/67/89/1011/1213/1415 Extended Control Registers
+ * are defined, and their bits definitions are as follows:
+ *
+ * Bit    Description
+ *  7     Reserved
+ *  6     EPPOINT5_ISO_ENABLE
+ *  5     EPPOINT5_SEND_STALL
+ *  4     EPPOINT5_OUT_DATA_SEQUENCE
+ *  3     Reserved
+ *  2     EPPOINT4_ISO_ENABLE
+ *  1     EPPOINT4_SEND_STALL
+ *  0     EPPOINT4_OUT_DATA_SEQUENCE
+ *
+ * Apparently, we can tell that the EP4 and EP5 share the same register, and
+ * the EP6 and EP7 share the same one, and the other EPs are defined in the
+ * same way.
+ *
+ * In the function usb_extend_ep_ctrl() we will obtain the mask/flag according
+ * to the bits definitions mentioned above. As for the control code, please
+ * refer to the definition of enum it82xx2_extend_ep_ctrl.
+ */
+static int usb_extend_ep_ctrl(uint8_t ep_idx, enum it82xx2_extend_ep_ctrl ctrl)
+{
+	uint8_t reg_idx, mask;
+	bool flag;
+
+	if (ep_idx < EP4) {
+		return -EINVAL;
+	}
+
+	if ((ctrl >= EXT_EP_DIR_IN) && (ctrl < EXT_EP_READY)) {
+	/* From EXT_EP_DIR_IN to EXT_EP_DISABLE */
+		reg_idx = ep_fifo_res[ep_idx % FIFO_NUM];
+		mask = 1 << (ext_ep_bit_shift[ep_idx - 4] * 2 + 1);
+		flag = epn_ext_ctrl_tbl[ctrl - EXT_EP_DIR_IN];
+		epn_ext_ctrl_cfg1(reg_idx, mask, flag);
+
+	} else if ((ctrl >= EXT_EP_ISO_DISABLE) && (ctrl < EXT_EP_DIR_IN)) {
+	 /* From EXT_EP_ISO_DISABLE to EXT_EP_DATA_SEQ_0 */
+		reg_idx = (ep_idx - 4) >> 1;
+		flag = (ep_idx % 2 == 0) ? (false) : (true);
```

Comment:
Is this equivalent to `flag = (ep_idx & 1);`?

Reply:
Yes, you're absolutely right. I should thought of it.Thanks for guidance, sir. I'll fix it in the next upload.

Reply:
Thanks for hint, sir.
Now I modify it to be 
`flag = !(ep_idx & 1) ? (false) : (true);`

Reply:
Why make it so complex? `flag = !!(ep_idx & 1);` should work just fine.

Reply:
Sorry that I misunderstood the question mark as a part of the ternary operator, now it's fixed as follows:

`flag = !!(ep_idx & 1);`


Since the flag is equal to the inverse of original (ep_idx % 2 == 0), so I swap the expr1 and expr2 when I use the flag as the condition as follows:

`mask = flag ? (ext_ctrl_tbl[ctrl] << 4) : (ext_ctrl_tbl[ctrl]);`

Thanks again for your great patience and helps, sir.